### PR TITLE
Retry count Metric

### PR DESCRIPTION
**Pull request by disa6302**
*Issue #, if available:*

*Description of changes:*
- Add retry count metric for every stream
- Add client level average retry count metric for all streams 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/client/src/StreamState.c [hunk 1]
```diff
@@ -138,6 +139,15 @@ STATUS defaultStreamStateTransitionHook(
             pKvsRetryStrategy->pRetryStrategy != NULL &&
             pKvsRetryStrategyCallbacks->executeRetryStrategyFn != NULL, STATUS_SUCCESS);
 
+    if(pKvsRetryStrategyCallbacks->getCurrentRetryAttemptNumberFn != NULL) {
+        if((countStatus = pKvsRetryStrategyCallbacks->getCurrentRetryAttemptNumberFn(pKvsRetryStrategy, &pKinesisVideoStream->diagnostics.streamApiCallRetryCount)) != STATUS_SUCCESS) {
+            DLOGW("Failed to get retry count. Error code: %08x", countStatus);
+        }
+        else {
+            DLOGD("Client state machine retry count: %llu", pKinesisVideoStream->diagnostics.streamApiCallRetryCount);
```
Comment (hassanctech) [inline src/client/src/StreamState.c:147]: UINT32 format specifier should be just `%lu`

**Review by hassanctech — COMMENTED** [review]
Some API results include caching.  Is this metric explicit in including actual API calls or just code paths that could make the API calls and if it's the latter then do we need to make some changes to make sure this is only incremented when the actual backend APIs are called instead of when it's possible a cached value is retrieved?

**Comment by disa6302** [comment]
> Some API results include caching. Is this metric explicit in including actual API calls or just code paths that could make the API calls and if it's the latter then do we need to make some changes to make sure this is only incremented when the actual backend APIs are called instead of when it's possible a cached value is retrieved?

It should be. This metric just reads what gets updated in exponential backoff. So if the count gets incremented in exponential backoff, the metric would reflect the same.

**Review by hassanctech — COMMENTED** [review]
Also, where is `streamApiCallRetryCount` actually incremented?

File: src/client/src/StreamState.c (hunk 1 above)
Comment (disa6302) [reply src/client/src/StreamState.c]: Oops. Fixed in new commit.

**Comment by disa6302** [comment]
> Also, where is `streamApiCallRetryCount` actually incremented?

This is where it gets incremented: https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/d10936c70ef00941226f431b9bda46eca5aa2f0a/src/utils/src/ExponentialBackoffRetryStrategy.c#L271 and `streamApiCallRetryCount` just gets updated with the value through the `getCurrentRetryAttemptNumberFn` callback.

File: src/client/src/Stream.c [hunk 2]
```diff
@@ -712,8 +712,14 @@ STATUS logStreamMetric(PKinesisVideoStream pKinesisVideoStream)
     // V2 stream information
     DLOGD("\tCurrent elementary frame rate (fps): %f ", streamMetrics.elementaryFrameRate);
 
+    // V3 stream information
+    DLOGD("\tAPI Call Retry Count : %llu", streamMetrics.streamApiCallRetryCount);
```
Comment (hassanctech) [inline src/client/src/Stream.c:716]: This one too `%lu`

File: src/client/src/Stream.c [hunk 3]
```diff
@@ -712,8 +712,14 @@ STATUS logStreamMetric(PKinesisVideoStream pKinesisVideoStream)
...
+
     // V1 client information
     DLOGD("\tTotal elementary frame rate (fps): %f ", clientMetrics.totalElementaryFrameRate);
```
Comment (hassanctech) [inline src/client/src/Stream.c:719]: `%lf` for clientMetrics.totalElementaryFrameRate as well since its also `DOUBLE`

**Review by hassanctech — COMMENTED** [review]
lgtm let's just wait for travis to finish

**Review by hassanctech — APPROVED** [review]
CI is clean, LGTM!

**Pull request merged**